### PR TITLE
Fix missing `_ZN6google8protobuf2io17SafeDoubleToFloatEd`

### LINKED ISSUE #190: Undefined symbol `_ZN6google8protobuf2io17SafeDoubleToFloatEd`
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

This may be an upstream issue but I want to check here first.

In downstream packages I'm seeing an undefined symbol error (e.g. in https://github.com/conda-forge/google-cloud-workflows-feedstock/pull/25).  I can reproduce this with both 4.23.1 and 4.23.2 releases:
```
$ mamba create -y -n test protobuf=4.23.2 libprotobuf=4.23.2
$ mamba activate test
$ python -c "from google.protobuf import descriptor"
/home/xylar/mambaforge/envs/test/lib/python3.11/site-packages/google/protobuf/internal/api_implementation.py:110: UserWarning: Selected implementation cpp is not available.
  warnings.warn(
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/xylar/mambaforge/envs/test/lib/python3.11/site-packages/google/protobuf/descriptor.py", line 51, in <module>
    from google.protobuf.pyext import _message
ImportError: /home/xylar/mambaforge/envs/test/lib/python3.11/site-packages/google/protobuf/pyext/_message.cpython-311-x86_64-linux-gnu.so: undefined symbol: _ZN6google8protobuf2io17SafeDoubleToFloatEd
```

### Installed packages

```shell
$ mamba list
# packages in environment at /home/xylar/mambaforge/envs/test:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
ca-certificates           2023.5.7             hbcca054_0    conda-forge
ld_impl_linux-64          2.40                 h41732ed_0    conda-forge
libabseil                 20230125.2      cxx17_h59595ed_2    conda-forge
libexpat                  2.5.0                hcb278e6_1    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 12.2.0              h65d4601_19    conda-forge
libgomp                   12.2.0              h65d4601_19    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libprotobuf               4.23.2               hd1fb520_1    conda-forge
libsqlite                 3.42.0               h2797004_0    conda-forge
libstdcxx-ng              12.2.0              h46fd767_19    conda-forge
libuuid                   2.38.1               h0b41bf4_0    conda-forge
libzlib                   1.2.13               h166bdaf_4    conda-forge
ncurses                   6.3                  h27087fc_1    conda-forge
openssl                   3.1.0                hd590300_3    conda-forge
pip                       23.1.2             pyhd8ed1ab_0    conda-forge
protobuf                  4.23.2          py311hbec7ed6_0    conda-forge
python                    3.11.3          h2755cc3_0_cpython    conda-forge
python_abi                3.11                    3_cp311    conda-forge
readline                  8.2                  h8228510_1    conda-forge
setuptools                67.7.2             pyhd8ed1ab_0    conda-forge
tk                        8.6.12               h27826a3_0    conda-forge
tzdata                    2023c                h71feb2d_0    conda-forge
wheel                     0.40.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
```

### Environment info

```shell
$ mamba info

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


     active environment : test
    active env location : /home/xylar/mambaforge/envs/test
            shell level : 2
       user config file : /home/xylar/.condarc
 populated config files : /home/xylar/mambaforge/.condarc
                          /home/xylar/.condarc
          conda version : 23.3.1
    conda-build version : 3.24.0
         python version : 3.10.11.final.0
       virtual packages : __archspec=1=x86_64
                          __glibc=2.35=0
                          __linux=5.19.0=0
                          __unix=0=0
       base environment : /home/xylar/mambaforge  (writable)
      conda av data dir : /home/xylar/mambaforge/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/xylar/mambaforge/pkgs
                          /home/xylar/.conda/pkgs
       envs directories : /home/xylar/mambaforge/envs
                          /home/xylar/.conda/envs
               platform : linux-64
             user-agent : conda/23.3.1 requests/2.29.0 CPython/3.10.11 Linux/5.19.0-42-generic ubuntu/22.04.2 glibc/2.35
                UID:GID : 1001:1001
             netrc file : None
           offline mode : False
```

### PR DESCRIPTION
**Pull request by h-vetinari**
~Check out what's going on in #190~

Closes #190

### CODE REVIEW

**Comment by xylar** [comment]
@h-vetinari, thanks for taking a look.

**Comment by h-vetinari** [comment]
I'm scratching my head why that symbol wouldn't be found (or not in the linked protobuf in the first place). If it's urgent we can switch from the CPP to the python builds, then at least things would run (if perhaps a bit slower)

**Comment by xylar** [comment]
@h-vetinari, absolutely no rush on my part.  I had just noticed this issue in a few different repos so I decided it was time to report it.  I think you're right that it would be better to take the time to debug the CPP build than to resort back to the python one.  I'm following along what you're trying in the hope that I can be more helpful at debugging these kinds of issues in the future.

**Comment by h-vetinari** [comment]
@coryan, if you have a moment, I feel I have need of your C++ foo here. I cannot figure out for the life of me why that symbol would not be found. To recap:
* The [source](https://github.com/protocolbuffers/protobuf/blob/v23.2/src/google/protobuf/io/strtod.cc) doesn't do anything unusual (well, `setup.py` doesn't correctly link abseil, but that's not it and I'm "fixing" this here already)
* That file definitely gets [included](https://github.com/protocolbuffers/protobuf/blob/v23.2/src/file_lists.cmake#L52) in the build of `libprotobuf`
* `libprotobuf` gets linked correctly(?) to `_message.cpython-PYVER-x86_64-linux-gnu.so`:
  ```
  INFO (protobuf,lib/python3.8/site-packages/google/protobuf/pyext/_message.cpython-38-x86_64-linux-gnu.so): Needed DSO lib/libprotobuf.so.23.2.0 found in conda-forge::libprotobuf-4.23.2-hd1fb520_1
  ```
* The symbol being asked for by `_message.cpython-PYVER-x86_64-linux-gnu.so` matches the signature of what's in libprotobuf
  ```
  >llvm-cxxfilt _ZN6google8protobuf2io17SafeDoubleToFloatEd
  google::protobuf::io::SafeDoubleToFloat(double)
  ```

**Comment by coryan** [comment]
> if you have a moment, I feel I have need of your C++ foo here.

Happy to help.  Don't take anything I say as gospel.  I am good at expressing myself with more confidence than I should.

> I cannot figure out for the life of me why that symbol would not be found.

Maybe this:

https://github.com/protocolbuffers/protobuf/blob/6609bea792af9624f460dad0213bb47fd1ccf623/cmake/libprotobuf.cmake#L42

In combination with a missing `PROTOBUF_EXPORT` here:

https://github.com/protocolbuffers/protobuf/blob/0ce1fce37f5663552448a35ef1cb84e28e764547/src/google/protobuf/io/strtod.h#L68

I do not know if it is *intentional* that this symbol is hidden, but at least on my build it is (the lowercase `t` as opposed to `T` is the giveaway):

```
nm /usr/local/lib64/libprotobuf.so.23.2.0 | c++filt | grep SafeDoubleToFloat
000000000018f350 t google::protobuf::io::SafeDoubleToFloat(double)
```

Contrast with an exported symbol:

```
nm /usr/local/lib64/libprotobuf.so.23.2.0 | c++filt | grep SimpleDtoa
000000000018f450 T google::protobuf::io::SimpleDtoa[abi:cxx11](double)
```

We need to ask the Protobuf team if they meant for these symbols to be used from outside the `.so`.


> To recap: The symbol in question
> 
> * The [source](https://github.com/protocolbuffers/protobuf/blob/v23.2/src/google/protobuf/io/strtod.cc) doesn't do anything unusual (well, `setup.py` doesn't correctly link abseil, but that's not it and I'm "fixing" this here already)
> * That file definitely gets [included](https://github.com/protocolbuffers/protobuf/blob/v23.2/src/file_lists.cmake#L52) in the build of `libprotobuf`
> * `libprotobuf` gets linked correctly(?) to `_message.cpython-PYVER-x86_64-linux-gnu.so`:
>   ```
>   INFO (protobuf,lib/python3.8/site-packages/google/protobuf/pyext/_message.cpython-38-x86_64-linux-gnu.so): Needed DSO lib/libprotobuf.so.23.2.0 found in conda-forge::libprotobuf-4.23.2-hd1fb520_1
>   ```
> * The symbol being asked for by `_message.cpython-PYVER-x86_64-linux-gnu.so` matches the signature of what's in libprotobuf
>   ```
>   >llvm-cxxfilt _ZN6google8protobuf2io17SafeDoubleToFloatEd
>   google::protobuf::io::SafeDoubleToFloat(double)
>   ```

That all seems correct, but for the hidden symbols flag.

**Comment by h-vetinari** [comment]
Ah, that's amazing, thanks!

> We need to ask the Protobuf team if they meant for these symbols to be used from outside the `.so`.

Well, it's their own python bindings that want to access it! 😅 

It seems they're not covering the build of the C++ backend against a shared `libprotobuf` (anymore? what with the move to upb) ... another issue to open?

**Comment by coryan** [comment]
> Ah, that's amazing, thanks!

No worries.

> > We need to ask the Protobuf team if they meant for these symbols to be used from outside the `.so`.
> 
> Well, it's their own python bindings that want to access it! 😅

HA!  Yeah, that kind of strongly suggests they should be public.

> It seems they're not covering the build of the C++ backend against a shared `libprotobuf` (anymore? what with the move to upb) ... another issue to open?

I created: https://github.com/protocolbuffers/protobuf/issues/12932

File: recipe/patches/0007-add-PROTOBUF_EXPORT-for-google-protobuf-io-SafeDoubl.patch [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+diff --git a/src/google/protobuf/io/strtod.h b/src/google/protobuf/io/strtod.h
+index 851c8e621..648f1ff1c 100644
+--- a/src/google/protobuf/io/strtod.h
++++ b/src/google/protobuf/io/strtod.h
+@@ -65,7 +65,7 @@ double NoLocaleStrtod(const char* str, char** endptr);
+ // Casts a double value to a float value. If the value is outside of the
+ // representable range of float, it will be converted to positive or negative
+ // infinity.
+-float SafeDoubleToFloat(double value);
++PROTOBUF_EXPORT float SafeDoubleToFloat(double value);
```
Comment (h-vetinari) [inline recipe/patches/0007-add-PROTOBUF_EXPORT-for-google-protobuf-io-SafeDoubl.patch:20]: Interestingly, adding `PROTOBUF_EXPORT` does not seem to be enough to make this symbol visible (even though other symbols in that TU already use that macro).

Comment (h-vetinari) [reply recipe/patches/0007-add-PROTOBUF_EXPORT-for-google-protobuf-io-SafeDoubl.patch:20]: Ah, it actually [needs](https://github.com/protocolbuffers/protobuf/blob/v23.2/src/google/protobuf/port_def.inc#L523-L526) `PROTOBUF_USE_DLLS` after all (which I find very confusing as an option on unix TBH...)

Comment (h-vetinari) [reply recipe/patches/0007-add-PROTOBUF_EXPORT-for-google-protobuf-io-SafeDoubl.patch:20]: Ah no, I need to actually rebuild the lib with the `PROTOBUF_EXPORT` patch... 🤦

**Comment by h-vetinari** [comment]
Thanks a lot for your help on this @coryan! 🙏

**Comment by xylar** [comment]
Thanks so much @h-vetinari, and @coryan as well.

**Pull request merged**